File: FBDeviceControl/Commands/FBDeviceEraseCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceEraseCommands.h"

#import <FBDeviceControl/FBDeviceControl-Swift.h>

#import "FBAMDevice.h"
#import "FBAMRestorableDeviceManager.h"
#import "FBDevice.h"

@interface FBDeviceEraseOperation : NSObject <FBiOSTargetSetDelegate>

@property (nonatomic, readonly, copy) NSString *udid;
@property (nonatomic, readonly, copy) NSString *ecid;
@property (nonatomic, readonly, assign) AMDCalls calls;
@property (nonatomic, readonly, strong) id<FBControlCoreLogger> logger;
@property (nonatomic, readonly, strong) FBAMRestorableDeviceManager *deviceManager;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNumber *> *eraseCallbackResult;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *deviceDetected;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *deviceWentAway;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *deviceCameBack;
@property (nonatomic, readonly, strong) dispatch_queue_t queue;

@end

static int const EraseCallbackValueGood = -10;
static NSTimeInterval const DetectTimeout = 10;
static NSTimeInterval const APICallbackTimeout = 15;
static NSTimeInterval const OfflineTimeout = 20;
static NSTimeInterval const OnlineTimeout = 300;

static int EraseCallback(NSString *identifier, int progress, void *context)
{
  FBDeviceEraseOperation *operation = (__bridge FBDeviceEraseOperation *)(context);
  [operation.logger log:[NSString stringWithFormat:@"Erase Callback is %d", progress]];
  [operation.eraseCallbackResult resolveWithResult:@(progress)];
  return 0;
}

@implementation FBDeviceEraseOperation

+ (instancetype)operationWithDevice:(FBDevice *)device logger:(id<FBControlCoreLogger>)logger
{
  dispatch_queue_t queue = dispatch_queue_create("com.facebook.fbdeviceerase", DISPATCH_QUEUE_SERIAL);
  return [[self alloc] initWithUDID:device.udid ecid:device.uniqueIdentifier calls:device.calls queue:queue logger:logger];
}

- (instancetype)initWithUDID:(NSString *)udid ecid:(NSString *)ecid calls:(AMDCalls)calls queue:(dispatch_queue_t)queue logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _udid = udid;
  _ecid = ecid;
  _logger = logger;
  _calls = calls;
  _queue = queue;
  _deviceManager = [[FBAMRestorableDeviceManager alloc] initWithCalls:calls workQueue:queue asyncQueue:queue ecidFilter:ecid logger:logger];
  _deviceManager.delegate = self;
  _eraseCallbackResult = FBMutableFuture.future;
  _deviceDetected = FBMutableFuture.future;
  _deviceWentAway = FBMutableFuture.future;
  _deviceCameBack = FBMutableFuture.future;

  return self;
}

#pragma mark Erase

- (FBFuture<NSNull *> *)erase
{
  FBAMRestorableDeviceManager *deviceManager = self.deviceManager;
  FBFuture<NSNull *> *deviceCameBack = self.deviceCameBack;
  FBFuture<NSNull *> *deviceWentAway = self.deviceWentAway;
  id<FBControlCoreLogger> logger = self.logger;
  return [[[[FBFuture
             onQueue:self.queue
             resolve:^FBFuture<NSNull *> * {
               NSError *error = nil;
               if (![deviceManager startListeningWithError:&error]) {
                 return [FBFuture futureWithError:error];
               }
               return [[self deviceDetected] timeout:DetectTimeout waitingFor:@"Device to be detected the first time"];
             }]
            onQueue:self.queue
            fmap:^FBFuture<NSNull *> *(id _) {
              [logger log:@"Device has been detected, starting erase API Call"];
              return [[self startErase] timeout:APICallbackTimeout waitingFor:@"Device erase API call to resolve"];
            }]
           onQueue:self.queue
           fmap:^FBFuture<NSNull *> *(NSNumber *eraseCallbackValueNumber) {
             const int eraseCallbackValue = eraseCallbackValueNumber.intValue;
             if (eraseCallbackValue != EraseCallbackValueGood) {
               return (FBFuture *)[[FBDeviceControlError
                                    describe:[NSString stringWithFormat:@"Erase callback was %d, not %d. Perhaps the device is not activated?", eraseCallbackValue, EraseCallbackValueGood]]
                                   failFuture];
             }
             [logger log:@"Device API call finished, waiting for device to go offline"];
             return [deviceWentAway timeout:OfflineTimeout waitingFor:@"Device to go offline"];
           }]
          onQueue:self.queue
          fmap:^FBFuture<NSNull *> *(id _) {
            [logger log:@"Device has gone offline, waiting for it to come back online"];
            return [deviceCameBack timeout:OnlineTimeout waitingFor:@"Device to come back"];
          }];
}

- (FBFuture<NSNumber *> *)startErase
{
  AMDCalls calls = self.calls;
  FBFuture<NSNumber *> *eraseCallbackResult = self.eraseCallbackResult;
  id<FBControlCoreLogger> logger = self.logger;
  NSString *udid = self.udid;
  return [FBFuture
          onQueue:self.queue
          resolve:^FBFuture<NSNumber *> * {
            calls.AMSInitialize(0);
            int status = calls.AMSEraseDevice((__bridge CFStringRef)(udid), EraseCallback, (__bridge void *)(self));
            [logger log:[NSString stringWithFormat:@"AMSEraseDevice had status %d", status]];
            return eraseCallbackResult;
          }];
}

#pragma mark FBiOSTargetSetDelegate

- (void)targetAdded:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{
  if (self.deviceDetected.state == FBFutureStateRunning) {
    [self.logger log:[NSString stringWithFormat:@"Got target %@ added for the first time", targetInfo]];
    [self.deviceDetected resolveWithResult:NSNull.null];
  } else {
    [self.logger log:[NSString stringWithFormat:@"Got target %@ added", targetInfo]];
    [self.deviceCameBack resolveWithResult:NSNull.null];
  }
}

- (void)targetRemoved:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{
  [self.logger log:[NSString stringWithFormat:@"Got target %@ removed", targetInfo]];
  [self.deviceWentAway resolveWithResult:NSNull.null];
}

- (void)targetUpdated:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{}

@end

@interface FBDeviceEraseCommands ()

@property (nonatomic, readonly, weak) FBDevice *device;
@property (nonatomic, readwrite, strong) FBAMRestorableDeviceManager *restorableDeviceManager;

@end

@implementation FBDeviceEraseCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBDevice *)target
{
  return [[self alloc] initWithDevice:target];
}

- (instancetype)initWithDevice:(FBDevice *)device
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;

  return self;
}

#pragma mark FBEraseCommands Implementation

- (FBFuture<NSNull *> *)erase
{
  id<FBControlCoreLogger> logger = [self.device.logger withName:[NSString stringWithFormat:@"erase_%@", self.device.udid]];
  return [[self.device
           activate]
          onQueue:self.device.workQueue
          fmap:^(id _) {
            FBDeviceEraseOperation *operation = [FBDeviceEraseOperation operationWithDevice:self.device logger:logger];
            return [[operation
                     erase]
                    onQueue:self.device.workQueue
                    doOnResolved:^(id __) {
                      [logger log:[NSString stringWithFormat:@"Device erase finished successfully %@", operation]];
                    }];
          }];
}

@end

File: FBDeviceControl/Commands/FBDeviceEraseCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

private let EraseCallbackValueGood: Int32 = -10
private let DetectTimeout: TimeInterval = 10
private let APICallbackTimeout: TimeInterval = 15
private let OfflineTimeout: TimeInterval = 20
private let OnlineTimeout: TimeInterval = 300

// MARK: - FBDeviceEraseOperation

private final class FBDeviceEraseOperation: NSObject, FBiOSTargetSetDelegate, @unchecked Sendable {

  private let udid: String
  private let calls: AMDCalls
  private let logger: any FBControlCoreLogger
  private let queue: DispatchQueue
  private let deviceManager: FBAMRestorableDeviceManager
  private let deviceDetected = FBMutableFuture<NSNull>()
  private let deviceWentAway = FBMutableFuture<NSNull>()
  private let deviceCameBack = FBMutableFuture<NSNull>()
  private let eraseCallbackResult = FBMutableFuture<NSNumber>()

  init(device: FBDevice, logger: any FBControlCoreLogger) {
    let queue = DispatchQueue(label: "com.facebook.fbdeviceerase")
    self.udid = device.udid
    self.calls = device.calls
    self.logger = logger
    self.queue = queue
    self.deviceManager = FBAMRestorableDeviceManager(
      calls: device.calls,
      work: queue,
      asyncQueue: queue,
      ecidFilter: device.uniqueIdentifier,
      logger: logger
    )
    super.init()
    self.deviceManager.delegate = self
  }

  // MARK: Erase

  func erase() async throws {
    try deviceManager.startListening()
    try await awaitEvent(deviceDetected, timeout: DetectTimeout, waitingFor: "Device to be detected the first time")
    logger.log("Device has been detected, starting erase API Call")
    let eraseCallbackValue = try await startErase()
    guard eraseCallbackValue == EraseCallbackValueGood else {
      throw FBDeviceControlError().describe("Erase callback was \(eraseCallbackValue), not \(EraseCallbackValueGood). Perhaps the device is not activated?").build()
    }
    logger.log("Device API call finished, waiting for device to go offline")
    try await awaitEvent(deviceWentAway, timeout: OfflineTimeout, waitingFor: "Device to go offline")
    logger.log("Device has gone offline, waiting for it to come back online")
    try await awaitEvent(deviceCameBack, timeout: OnlineTimeout, waitingFor: "Device to come back")
  }

  // MARK: Private

  /// Issues `AMSEraseDevice` on the serial queue and awaits the value delivered by
  /// the C erase callback (with a timeout).
  private func startErase() async throws -> Int32 {
    queue.async { [self] in
      let selfContext = Unmanaged.passUnretained(self).toOpaque()
      _ = calls.AMSInitialize?(0)
      let status =
        calls.AMSEraseDevice?(
          udid as CFString,
          { _, progress, context in
            guard let context else {
              return 0
            }
            let operation = Unmanaged<FBDeviceEraseOperation>.fromOpaque(context).takeUnretainedValue()
            operation.logger.log("Erase Callback is \(progress)")
            operation.eraseCallbackResult.resolve(withResult: NSNumber(value: progress))
            return 0
          },
          selfContext
        ) ?? -1
      logger.log("AMSEraseDevice had status \(status)")
    }
    // swiftlint:disable:next force_cast
    let timed = convertFBMutableFuture(eraseCallbackResult).timeout(APICallbackTimeout, waitingFor: "Device erase API call to resolve") as! FBFuture<NSNumber>
    let result = try await bridgeFBFuture(timed)
    return result.int32Value
  }

  private func awaitEvent(_ future: FBMutableFuture<NSNull>, timeout: TimeInterval, waitingFor description: String) async throws {
    // swiftlint:disable:next force_cast
    let timed = convertFBMutableFuture(future).timeout(timeout, waitingFor: description) as! FBFuture<NSNull>
    try await bridgeFBFutureVoid(timed)
  }

  // MARK: FBiOSTargetSetDelegate

  func targetAdded(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {
    if deviceDetected.state == .running {
      logger.log("Got target \(targetInfo) added for the first time")
      deviceDetected.resolve(withResult: NSNull())
    } else {
      logger.log("Got target \(targetInfo) added")
      deviceCameBack.resolve(withResult: NSNull())
    }
  }

  func targetRemoved(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {
    logger.log("Got target \(targetInfo) removed")
    deviceWentAway.resolve(withResult: NSNull())
  }

  func targetUpdated(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {}
}

// MARK: - FBDeviceEraseCommands

public final class FBDeviceEraseCommands: NSObject, FBiOSTargetCommand, EraseCommands {

  private weak var device: FBDevice?

  public class func commands(with target: any FBiOSTarget) -> Self {
    // swiftlint:disable:next force_cast
    unsafeDowncast(FBDeviceEraseCommands(device: target as! FBDevice), to: self)
  }

  init(device: FBDevice) {
    self.device = device
    super.init()
  }

  // MARK: EraseCommands

  public func erase() async throws {
    guard let device else {
      throw FBDeviceControlError().describe("Device is nil").build()
    }
    let logger = device.logger?.withName("erase_\(device.udid)") ?? FBControlCoreGlobalConfiguration.defaultLogger
    try await device.activate()
    let operation = FBDeviceEraseOperation(device: device, logger: logger)
    try await operation.erase()
    logger.log("Device erase finished successfully \(operation)")
  }
}

// MARK: - FBDevice+EraseCommands

extension FBDevice: EraseCommands {

  public func erase() async throws {
    try await eraseCommands().erase()
  }
}
